processor/processor.go

```diff
@@ -18,21 +18,52 @@
 package processor
 
 import (
-	"github.com/elastic/apm-server/config"
-	"github.com/elastic/beats/libbeat/beat"
+	"github.com/santhosh-tekuri/jsonschema"
+
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/validation"
+	"github.com/elastic/beats/libbeat/monitoring"
 )
 
 type NewProcessor func() Processor
 
 type Processor interface {
 	Validate(map[string]interface{}) error
-	Decode(map[string]interface{}) (Payload, error)
+	Decode(map[string]interface{}) (model.Payload, error)
 	Name() string
 }
-type Payload interface {
-	Transform(config.Config) []beat.Event
+
+type PayloadDecoder func(map[string]interface{}) (model.Payload, error)
+
+type PayloadProcessor struct {
```

Comment:
I find this naming very confusing. 
Now you have a `Processor` interface, that is implemented by one type `PayloadProcessor`. That actually returns a `payload` for the decoding part, instead of acting on a `payload`. Then the actual `model.payload` types again do some decoding and validation.
I suggest to avoid having a name `Payload` in the `processor` package.

Reply:
I can see why this would be confusing. It's partly due to the fact that it's been split out from the mega PR initially had.

I'll share my mental model to see if that helps:
"Processing" to me includes taking the raw data and _validating_ it, _decoding_ it into a go structure and eventually _transformining_ it into an event (transformation happens by way of a method on the go struct, so it's not technically defined by the processor, but a separate refactor can change this if we want). In intake v1, the APM Server deals only with Payloads. For intake v2, we'll also have to deal with individual objects/models. 

There's a general interface `Processor` today is only implemented by the `PayloadProcessor`s. In the future we would have a `ModelProcessor` that would know how to take individual data pieces and _validate_, _decode_ and _transform_ them. 

Let me know if it makes more sense in this context.

Reply:
Thanks for explaining. I think I am clear on the context. I still think that this is very confusing to call this `PayloadProcessor` as it doesn't operate on what is called a payload further on. 
I'd rather call it something like `ProcessorImpl` than adding `Payload`.

Reply:
We discussed offline and decided to keep as is for now given that subsequent PRs will address this.